torch/distributed/fsdp/_optim_utils.py

```diff
@@ -1427,9 +1427,26 @@ def _unflatten_orig_param_states(
     fsdp_state = fsdp_param_info.state
     for fqn, gathered_state in output_states.items():
         value = gathered_state[state_name]
-
         param_idx = fsdp_param_info.param_indices[fqn]
-        value = value.reshape(flat_param._shapes[param_idx])
+
+        if isinstance(value, DTensor):
+            placement = value.placements[0]
+            # If gathered state is a DTensor and its TP placement is not Replicate(), we need to
+            # gather the tensor on its TP dimension before chunking them into DTensor again.
+            if placement != Replicate():
```

Comment:
Let's add a TODO here to mention this solution is not general and is only applicable for PTD TP.